task/task.go

```diff
@@ -0,0 +1,155 @@
+}
+
+// task is a basic implementation for a `Task`.
+type task struct {
+	// The reducer for message handling logic.
+	reducer Reducer
+
+	// The buffered channel that incoming messages are written to.
+	input chan messageWithResponder
+}
+
+// New returns a new task with the given reducer and buffer capacity. The
+// buffer capacity is the number of messages that can be buffered for
+// processing before the task can no longer accept more messages (until space
+// in the buffer is freed up by processing messages in the buffer).
+func New(reducer Reducer, cap int) Task {
+	return &task{
+		reducer: reducer,
+		input:   make(chan messageWithResponder, cap),
+	}
+}
+
+// Run implements the `Runner` interface (in order to implement the `Task`
+// interface). This function blocks. The task will continue to run until it is
+// signalled to terminate by the context.
+func (task *task) Run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case message := <-task.input:
+			message.responder <- task.reduce(flatten(message.message))
+			close(message.responder)
+		}
+	}
+}
+
+// Send implements the `Sender` interface (in order to implement the `Task`
+// interface). It returns a channel to which the (possibly nil) response will
+// be written, and a bool indicating whether the message was able to be sent;
+// true indicates the message was sent, and false indicates that the task
+// currently has a full buffer and won't accept the message. In the latter case
+// the returned channel will be nil.
+func (task *task) Send(message Message) (<-chan Messages, bool) {
+	responder := make(chan Messages, 1)
+	m := messageWithResponder{message: message, responder: responder}
+	select {
+	case task.input <- m:
+		return responder, true
+	default:
+		return nil, false
+	}
+}
+
+// reduce will handle the reduction of a given message for a task. It is
+// assumed that the message is flattened. It will always return a `Messages`
+// type which contains the responses of the reducer for the given message(s).
+// The returned `Messages` is flattened.
+func (task *task) reduce(message Message) Messages {
+	messages := Messages{}
+	switch message := message.(type) {
+	case Messages:
+		for _, msg := range message {
+			messages = append(messages, task.reducer.Reduce(task, msg))
+		}
+	default:
+		messages = append(messages, task.reducer.Reduce(task, message))
+	}
+	return flatten(messages).(Messages)
+}
+
+// flatten takes a message and effectively flattens it out to depth 1, where
+// the depth is determined by the level of nested Messages. If `flatten`
+// receives a message that is not a `Messages`, it will return the same
+// message. Otherwise, it will return a `Messages` type where the internal list
+// of messages contain no `Messages`; these will be flattened out.
+func flatten(message Message) Message {
+	switch message := message.(type) {
+	case Messages:
+		msgs := Messages{}
+		for _, msg := range message {
+			m := flatten(msg)
```

Comment:
It might make more sense to have the type signature `flatten(ms Messages) Messages` instead of having the top-level type switch since everywhere that `flatten` is called you either (a) know that the message type is `Messages`, or (b) do a type switch on the return value.

Reply:
Keeping as is for now but we can consider this change again at a later stage if necessary.